File: csharp/src/Noie.Nds.Api/Adapter/ResultAdapter.cs
using Noie.Nds.Api.Result;

namespace Noie.Nds.Api.Adapter;

/// <summary>
/// [Index: NDS-CSHARP-RESULTADAPTER-000] Result adapter (NdsResult/NdsError ↔ proto-compatible shapes).
/// 
/// <para>[Semantic] Conversion utilities for structured success/failure envelopes.</para>
/// </summary>
public static class ResultAdapter
{
    /// <summary>
    /// [Index: NDS-CSHARP-RESULTADAPTER-010] Create a proto-compatible result tuple.
    /// </summary>
    public static (bool Success, string? ErrorCode, string? ErrorMessage, Dictionary<string, string>? Details) ToProtoData<T>(NdsResult<T> result)
    {
        if (result.IsSuccess)
        {
            return (true, null, null, null);
        }
        
        var details = new Dictionary<string, string>();
        foreach (var kvp in result.Error.Details)
        {
            details[kvp.Key] = kvp.Value?.ToString() ?? "";
        }
        
        return (false, result.Error.Code, result.Error.Message, details);
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-RESULTADAPTER-020] Create a proto-compatible error tuple.
    /// </summary>
    public static (string Code, string Message, Dictionary<string, string> Details) ToProtoData(NdsError error)
    {
        var details = new Dictionary<string, string>();
        foreach (var kvp in error.Details)
        {
            details[kvp.Key] = kvp.Value?.ToString() ?? "";
        }
        
        return (error.Code, error.Message, details);
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-RESULTADAPTER-021] Create a domain error from proto-compatible data.
    /// </summary>
    public static NdsError? FromProtoData(string? code, string? message, IReadOnlyDictionary<string, string>? details)
    {
        if (string.IsNullOrEmpty(code)) return null;
        
        var objectDetails = new Dictionary<string, object>();
        if (details != null)
        {
            foreach (var kvp in details)
            {
                objectDetails[kvp.Key] = kvp.Value;
            }
        }
        
        return NdsError.Of(code, message ?? "", objectDetails);
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-RESULTADAPTER-022] Create a domain result from proto-compatible data.
    /// </summary>
    public static NdsResult<T> FromProtoData<T>(bool success, T? data, string? errorCode, string? errorMessage, IReadOnlyDictionary<string, string>? errorDetails)
    {
        if (success && data != null)
        {
            return NdsResult<T>.Success(data);
        }
        
        var error = FromProtoData(errorCode, errorMessage, errorDetails);
        return NdsResult<T>.Failure(error ?? NdsError.Of("UNKNOWN", "Unknown error"));
    }
}

File: java/src/main/java/noie/linmimeng/noiedigitalsystem/api/adapter/ResultAdapter.java
package noie.linmimeng.noiedigitalsystem.api.adapter;

import noie.linmimeng.noiedigitalsystem.api.result.NdsResult;
import noie.linmimeng.noiedigitalsystem.api.result.NdsError;
import java.util.Map;
import java.util.HashMap;

/**
 * [Index] NDS-JAVA-RESULT-000
 * [Semantic] NdsResult / NdsError Domain ↔ Proto conversion.
 *
 * @since 2.0.0
 */
public final class ResultAdapter {
    
    private ResultAdapter() {
        // [Index] NDS-JAVA-RESULT-001 [Constraint] Utility class; instantiation is prohibited.
    }
    
    // ========================================================================
    // [Index] NDS-JAVA-RESULT-010 [Semantic] NdsResult conversion.
    // ========================================================================
    
    /**
     * @param domain domain NdsResult (nullable)
     * @return Proto NdsResult; null if input is null
     */
    public static noie.linmimeng.noiedigitalsystem.api.proto.common.NdsResult toProto(NdsResult<?> domain) {
        if (domain == null) {
            return null;
        }
        
        noie.linmimeng.noiedigitalsystem.api.proto.common.NdsResult.Builder builder = 
            noie.linmimeng.noiedigitalsystem.api.proto.common.NdsResult.newBuilder()
                .setSuccess(domain.isSuccess());
        
        if (!domain.isSuccess() && domain.error() != null) {
            builder.setError(toProto(domain.error()));
        }
        
        return builder.build();
    }
    
    /** @return Proto NdsResult with success=true */
    public static noie.linmimeng.noiedigitalsystem.api.proto.common.NdsResult successProto() {
        return noie.linmimeng.noiedigitalsystem.api.proto.common.NdsResult.newBuilder()
            .setSuccess(true)
            .build();
    }
    
    /**
     * @param code error code
     * @param message error message
     * @return Proto NdsResult with success=false and embedded error
     */
    public static noie.linmimeng.noiedigitalsystem.api.proto.common.NdsResult failureProto(String code, String message) {
        return noie.linmimeng.noiedigitalsystem.api.proto.common.NdsResult.newBuilder()
            .setSuccess(false)
            .setError(noie.linmimeng.noiedigitalsystem.api.proto.common.NdsError.newBuilder()
                .setCode(code)
                .setMessage(message)
                .build())
            .build();
    }
    
    // ========================================================================
    // [Index] NDS-JAVA-RESULT-020 [Semantic] NdsError conversion.
    // ========================================================================
    
    /**
     * @param domain domain NdsError (nullable)
     * @return Proto NdsError; null if input is null
     */
    public static noie.linmimeng.noiedigitalsystem.api.proto.common.NdsError toProto(NdsError domain) {
        if (domain == null) {
            return null;
        }
        
        noie.linmimeng.noiedigitalsystem.api.proto.common.NdsError.Builder builder = 
            noie.linmimeng.noiedigitalsystem.api.proto.common.NdsError.newBuilder()
                .setCode(domain.code())
                .setMessage(domain.message());
        
        if (domain.details() != null) {
            // [Index] NDS-JAVA-RESULT-021 [Behavior] Normalize Map<String, Object> into Map<String, String>.
            Map<String, String> stringDetails = new HashMap<>();
            for (Map.Entry<String, Object> entry : domain.details().entrySet()) {
                stringDetails.put(entry.getKey(), String.valueOf(entry.getValue()));
            }
            builder.putAllDetails(stringDetails);
        }
        
        return builder.build();
    }
    
    /**
     * @param proto Proto NdsError (nullable)
     * @return domain NdsError; null if proto is null or code is empty
     */
    public static NdsError fromProto(noie.linmimeng.noiedigitalsystem.api.proto.common.NdsError proto) {
        if (proto == null || proto.getCode().isEmpty()) {
            return null;
        }
        
        Map<String, Object> details = new HashMap<>(proto.getDetailsMap());
        return NdsError.of(proto.getCode(), proto.getMessage(), details);
    }
}
